wiring/src/spark_wiring_ble.cpp

```diff
@@ -1549,15 +1468,35 @@ bool BlePeerDevice::getServiceByUUID(BleService& service, const BleUuid& uuid) c
     return false;
 }
 
-Vector<BleCharacteristic> BlePeerDevice::characteristics() {
+Vector<BleCharacteristic> BlePeerDevice::characteristics() const {
     return impl()->characteristics();
 }
 
-size_t BlePeerDevice::characteristics(BleCharacteristic* chars, size_t count) {
-    CHECK_TRUE(chars && count > 0, SYSTEM_ERROR_INVALID_ARGUMENT);
+Vector<BleCharacteristic> BlePeerDevice::characteristics(const BleService& service) const {
+    Vector<BleCharacteristic> characteristics;
+    for (const auto& characteristic : impl()->characteristics()) {
+        if (characteristic.impl()->svcUUID() == service.impl()->UUID()) {
```

Comment:
This checks if a characteristic has a Service UUID that is the same as the one we're looking for. If a GATT server has multiple services with the same UUID, then this would return characteristics from multiple Services, instead of just the Service we're looking for.

Reply:
Technically, we should not add more than two services with the same UUID for a BLE peripheral. UUID is shorted for Universal Unique Identifier and BLE introduces it, so we should comply with the way we use it. We can absolutely fix it by checking the attribute handle range, but I'm wondering if it is worthy for us to introduce a wiring API that returns a vector of services with the same UUID for particular and abnormal cases.

Reply:
Unique means that it identifies a unique service, but there can be multiple copies of the same service on a device (for example, multiple Battery Services with UUID 0x180F on a device with more than one battery).

Here's from the Bluetooth specification:
"A device or higher level specification may have multiple service definitions and may have multiple service definitions with the same service UUID."

From Core specification v5.2, Volume 3, Part G, section 3.1

Reply:
👍 That makes perfect sense! Thanks for diving into the specification and pointing it out! I believe that we are on the same page now. I will make proper changes later on.